### PR TITLE
feat: add StateProviderFactor::state_by_block_hash

### LINKED ISSUE #2387: Add additional StateProviderFactory functions
### Describe the feature

Now that we have a wrapper type for:

https://github.com/paradigmxyz/reth/blob/26fe9cc10f9866fbe000ace905f711220d5cdacb/crates/storage/provider/src/providers/mod.rs#L39-L44

we only can request the state for historic state or the pending block, for block building however it could be the case that the block can be another block in the tree.

## TODO
* add additional function(s), like `find_state_by_block_hash` or `pending_or_historic_state_by_block_hash`

ref #2368

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mattsse**
Closes #2387

this clarifies a few things and adds a new function that checks tree+db for state

### CODE REVIEW

**Review by rkrasiuk — APPROVED** [review]
lgtm

File: crates/storage/provider/src/providers/mod.rs [hunk 1]
```diff
@@ -270,6 +270,16 @@ where
         self.database.history_by_block_hash(block_hash)
     }
 
+    fn state_by_block_hash(&self, block: BlockHash) -> Result<StateProviderBox<'_>> {
+        // check tree first
```
Comment (gakonst) [inline crates/storage/provider/src/providers/mod.rs:274]: Should this also take the txpool into account? e.g. what does eth_storageAt  with tag = `Pending` use?

File: crates/storage/provider/src/traits/state.rs [hunk 2]
```diff
@@ -72,7 +72,18 @@ pub trait StateProvider: BlockHashProvider + AccountProvider + Send + Sync {
...
-/// `BlockNumber` or the latest state.
+/// `BlockNumber`, the latest state, or the pending state.
+///
+/// This type differentiates states into `historical`, `latest` and `pending`, where the `latest`
+/// block determines what is historical or pending: `[historical..latest..pending]`.
+///
+/// The `latest` state represents the state after the most recent block has been committed to the
+/// database, `historical` states are states that have been committed to the database before the
+/// `latest` state, and `pending` states are states that have not yet been committed to the
+/// database which may or may not become the `latest` state, depending on consensus.
```
Comment (gakonst) [inline crates/storage/provider/src/traits/state.rs:83]: pending may also refer to state including txpool? e.g. for nonces?

File: crates/storage/provider/src/traits/state.rs [hunk 3]
```diff
@@ -104,13 +115,26 @@ pub trait StateProviderFactory: Send + Sync {
...
+    /// Note: this only looks at historical blocks, not pending blocks.
     fn history_by_block_number(&self, block: BlockNumber) -> Result<StateProviderBox<'_>>;
 
-    /// Returns a [StateProvider] indexed by the given block hash.
+    /// Returns a historical [StateProvider] indexed by the given block hash.
+    ///
+    /// Note: this only looks at historical blocks, not pending blocks.
     fn history_by_block_hash(&self, block: BlockHash) -> Result<StateProviderBox<'_>>;
 
+    /// Returns _any_[StateProvider] with matching block hash.
```
Comment (gakonst) [inline crates/storage/provider/src/traits/state.rs:129]:
```suggestion
    /// Returns _any_ [StateProvider] with matching block hash.
```

**Review by gakonst — APPROVED** [review]

**Pull request merged**